### PR TITLE
Add json->sql deserialisation and execution.

### PR DESCRIPTION
**Pull request by Maxxen**
This is a continuation of #6647 and  adds three new functions:
- `json_deserialize_sql(JSON)`: convert a json serialized sql select statement back into an (equivalent) sql statement string.
- `json_execute_serialized_sql(JSON)`: execute a json serialized sql select statement directly as a table function - basically as if it were a subquery.
- `PRAGMA json_execute_serialized_sql(VARCHAR)`: execute a json serialized sql select statement as if it were a top level statement through a pragma instead. Basically deserialises the given json back into a sql string which is then executed normally.

I have tested this by hacking in a json ser/de roundtrip in the DeserializedStatementVerifier, and it passes the entire duckdb unit test suite, so I am reasonably confident that this implementation is correct and serializes e.g. all the types, expressions and values correctly, but I will add a proper deserialization/serialization verification pass when I add the binary FormatDe/Serializer in core duckdb.

### CODE REVIEW

File: src/common/types/vector.cpp [hunk 1]
```diff
@@ -964,66 +964,85 @@ void Vector::Serialize(idx_t count, Serializer &serializer) {
...
 			Flatten(count);
 			auto &entries = StructVector::GetEntries(*this);
+
+			// Serialize entries as a list
+			serializer.SetTag("children");
+			serializer.OnListBegin(entries.size());
 			for (auto &entry : entries) {
+				serializer.OnObjectBegin();
 				entry->FormatSerialize(serializer, count);
+				serializer.OnObjectBegin();
```
Comment (Mytherin) [inline src/common/types/vector.cpp:1014]: Should this be `OnObjectEnd`?

File: src/common/types/vector.cpp [hunk 2]
```diff
@@ -964,66 +964,85 @@ void Vector::Serialize(idx_t count, Serializer &serializer) {
...
+				serializer.WriteProperty("offset", entries[i].offset);
+				serializer.WriteProperty("length", entries[i].length);
+				serializer.OnObjectEnd();
+			}
+			serializer.OnListEnd(count);
+			serializer.SetTag("child");
+			serializer.OnObjectBegin();
 			child.FormatSerialize(serializer, list_size);
+			serializer.OnObjectBegin();
```
Comment (Mytherin) [inline src/common/types/vector.cpp:1045]: Same here - should this be `OnObjectEnd`?

File: src/parser/tableref/pivotref.cpp [hunk 3]
```diff
@@ -371,3 +371,73 @@ unique_ptr<TableRef> PivotRef::FormatDeserialize(FormatDeserializer &source) {
 }
 
 } // namespace duckdb
+
+/*
+ json_serialize_sql('SELECT id, MIN(COLUMNS(''index[0-9]'')) FROM grouped_table GROUP BY all;', format := CAST('t' AS
```
Comment (Mytherin) [inline src/parser/tableref/pivotref.cpp:376]: Left-over debug comment?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good. Some minor comments below.

**Comment by Maxxen** [comment]
yo-yo @Mytherin

**Comment by Mytherin** [comment]
Thanks! LGTM. Next up replacing the existing (de)serialize methods nad unifying all the (de)serialize infrastructure :)

File: extension/json/json_deserializer.cpp [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+		// unreachable?
+		throw InternalException("Cannot get value from non-array/object");
+	}
+	return val;
+}
+
+void JsonDeserializer::ThrowTypeError(yyjson_val *val, const char *expected) {
+	auto actual = yyjson_get_type_desc(val);
+	auto &parent = Current();
+	if (yyjson_is_obj(parent.val)) {
```
Comment (Wang-Ji20) [inline extension/json/json_deserializer.cpp:43]: These two branches don't throw anything, and the string msg is discarded right away. Is that a bug?

Comment (Maxxen) [reply extension/json/json_deserializer.cpp:43]: Yes, nice catch! I've pushed a fix in #8291

**Pull request merged**